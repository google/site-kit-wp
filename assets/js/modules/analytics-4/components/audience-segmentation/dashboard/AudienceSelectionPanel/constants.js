/**
 * AudienceSelectionPanel components - constants.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const AUDIENCE_ADD_GROUP_NOTICE_SLUG =
	'audience-segmentation-add-group-notice';
export const AUDIENCE_SELECTION_PANEL_OPENED_KEY =
	'googlesitekit-audience-selection-panel-opened';
export const AUDIENCE_SELECTION_FORM = 'audience-selection-form';
export const AUDIENCE_SELECTED = 'audience-selected';
export const AUDIENCE_SELECTION_CHANGED = 'audience-selection-changed';
export const AUDIENCE_CREATION_NOTICE_SLUG =
	'audience-segmentation-creation-notice';
export const AUDIENCE_CREATION_SUCCESS_NOTICE_SLUG =
	'audience-segmentation-creation-success-notice';
export const MIN_SELECTED_AUDIENCES_COUNT = 1;
export const MAX_SELECTED_AUDIENCES_COUNT = 3;
export const AUDIENCE_CREATION_EDIT_SCOPE_NOTICE_SLUG =
	'audience-creation-edit-scope-notice';
export const AUDIENCE_CREATION_FORM = 'audience-creation-form';
