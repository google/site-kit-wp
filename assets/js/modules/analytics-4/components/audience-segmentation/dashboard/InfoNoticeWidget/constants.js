/**
 * Constants related to audience segmentation info notices.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

/**
 * UI key to display audience segmentation info notice.
 *
 * @since 1.135.0
 * @private
 */
export const AUDIENCE_INFO_NOTICE_HIDE_UI =
	'audience-segmentation-info-notice-ui';

/**
 * Slug for audience segmentation info notice.
 *
 * @since 1.130.0
 * @private
 */
export const AUDIENCE_INFO_NOTICE_SLUG = 'audience-segmentation-info-notice';

// Notice slugs: new-visitors, compare-metrics, custom-audiences, purchasers, returning-visitors, compare-new-returning, or compare-cities

/**
 * List of info notices for audiences.
 *
 * @since 1.130.0
 * @private
 */
export const AUDIENCE_INFO_NOTICES = [
	{
		slug: 'new-visitors',
		content: __(
			'The higher the portion of new visitors you have, the more your audience is growing. Looking at what content brings them to your site may give you insights on how to reach even more people.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-metrics',
		content: __(
			'Select up to three visitor groups to display on the dashboard and easily compare metrics between them.',
			'google-site-kit'
		),
	},
	{
		slug: 'custom-audiences',
		content: __(
			'Configure your own custom audiences in Analytics to gain deeper insights into visitor behavior, for example consider creating an “Existing customers” or “Subscribers” segment, depending on what goals you have for your site.',
			'google-site-kit'
		),
	},
	{
		slug: 'purchasers',
		content: __(
			'Select the Purchasers visitor group to gain insights into which visitors bring the most revenue to your site.',
			'google-site-kit'
		),
	},
	{
		slug: 'returning-visitors',
		content: __(
			'The more returning visitors your site has, the stronger and more loyal an audience you’re building. Check which content brings people back to your site - it might help you create a strategy to build a community.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-new-returning',
		content: __(
			'Compare the ratio of “new” to “returning” visitors – this can give you insights on whether you have more people stopping by as a one-off, or more loyal visitors.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-cities',
		content: __(
			'Check the cities which bring you more new vs more returning visitors – there might be new audiences you could engage with in locations you hadn’t thought about.',
			'google-site-kit'
		),
	},
];
