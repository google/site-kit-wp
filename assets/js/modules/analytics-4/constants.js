/**
 * GA4 module constants.
 *
 * Site Kit by Google, Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const ACTIVATION_STEP_REMINDER = 0;
export const ACTIVATION_STEP_SETUP = 1;
export const ACTIVATION_STEP_SUCCESS = 2;
export const ACTIVATION_STEP_IN_PROGRESS = 3;

export const ACTIVATION_ACKNOWLEDGEMENT_TOOLTIP_STATE_KEY =
	'activation-acknowledgement-tooltip-state';

export const GA4_ACTIVATION_BANNER_STATE_KEY = 'ga4-activation-banner-state';

export const ENHANCED_MEASUREMENT_ACTIVATION_BANNER_TOOLTIP_STATE_KEY =
	'enhanced-measurement-activation-banner-tooltip-state';
export const LEGACY_ENHANCED_MEASUREMENT_ACTIVATION_BANNER_DISMISSED_ITEM_KEY =
	'enhanced-measurement-activation-banner-dismissed-item';

export const REPORT_ARGS_SELECTED_METRIC_KEY = '_r.explorerCard..selmet';
export const REPORT_ARGS_SELECTED_DIMENSION_KEY = '_r.explorerCard..seldim';
export const REPORT_ARGS_DATA_FILTERS_KEY = '_r..dataFilters';
export const REPORT_ARGS_NAV_KEY = '_r..nav';

export const KM_CONNECT_GA4_CTA_WIDGET_DISMISSED_ITEM_KEY =
	'key-metrics-connect-ga4-cta-widget';

/**
 * Date that Site Kit will automatically switch to using GA4: September 25, 2023.
 *
 * With Site Kit's maximum date range being three months, once this period has passed since the UA cutoff date,
 * it will no longer be able to show any historical data for UA and will automatically switch to GA4.
 * For operational reasons, we automatically make the switch a few days before the three month cutoff point.
 *
 * @since 1.107.0
 * @see UA_CUTOFF_DATE.
 */
export const GA4_AUTO_SWITCH_DATE = '2023-09-25';
export const MODULE_SLUG_ANALYTICS_4 = 'analytics-4';
