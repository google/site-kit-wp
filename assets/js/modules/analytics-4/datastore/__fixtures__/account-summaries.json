{
  "accountSummaries": [
    {
      "account": "accounts/123456789",
      "displayName": "Example Org",
      "name": "accountSummaries/123456789",
      "_id": "123456789",
      "propertySummaries": []
    },
    {
      "account": "accounts/987654321",
      "displayName": "Example Com",
      "name": "accountSummaries/987654321",
      "propertySummaries": [
        {
          "displayName": "Example Property",
          "parent": "accounts/987654321",
          "property": "properties/1122334455",
          "_id": "1122334455",
          "_accountID": "987654321"
        }
      ],
      "_id": "987654321"
    },
    {
      "account": "accounts/152925174",
      "displayName": "Example Net",
      "name": "accountSummaries/152925174",
      "propertySummaries": [
        {
          "displayName": "Example Property",
          "parent": "accounts/152925174",
          "property": "properties/1122334455",
          "_id": "1122334455",
          "_accountID": "152925174"
        }
      ],
      "_id": "152925174"
    }
  ],
  "nextPageToken": null
}
