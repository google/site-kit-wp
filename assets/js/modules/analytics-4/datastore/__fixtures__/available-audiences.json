[
    {
        "name": "properties/12345/audiences/1",
        "description": "All users",
        "displayName": "All visitors",
        "audienceType": "DEFAULT_AUDIENCE",
        "audienceSlug": "all-users"
    },
    {
        "name": "properties/12345/audiences/2",
        "description": "Users who have made a purchase",
        "displayName": "Purchasers",
        "audienceType": "DEFAULT_AUDIENCE",
        "audienceSlug": "purchasers"
    },
    {
        "name": "properties/12345/audiences/3",
        "description": "People who visited the site for the first time",
        "displayName": "New visitors",
        "audienceType": "SITE_KIT_AUDIENCE",
        "audienceSlug": "new-visitors"
    },
    {
        "name": "properties/12345/audiences/4",
        "description": "People who have visited your site at least once before",
        "displayName": "Returning visitors",
        "audienceType": "SITE_KIT_AUDIENCE",
        "audienceSlug": "returning-visitors"
    },
    {
        "name": "properties/12345/audiences/5",
        "description": "Description",
        "displayName": "Test Audience",
        "audienceType": "USER_AUDIENCE",
        "audienceSlug": ""
    }
]
