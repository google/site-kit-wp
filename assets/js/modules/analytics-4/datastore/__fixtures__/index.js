/**
 * GA4 Datastore Fixtures.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { default as accountSummaries } from './account-summaries.json';
export { default as audiences } from './audiences.json';
export { default as availableAudiences } from './available-audiences.json';
export { default as container } from './container.json';
export { default as containerE2E } from './container-e2e.json';
export { default as containerDestinations } from './container-destinations.json';
export { default as conversionEvents } from './conversion-events.json';
export { default as createProperty } from './create-property.json';
export { default as createWebDataStream } from './create-webdatastream.json';
export { default as defaultSettings } from './default-settings.json';
export { default as defaultEnhancedMeasurementSettings } from './default-enhanced-measurement-settings.json';
export { default as pageTitles } from './page-titles.json';
export { default as properties } from './properties.json';
export { default as report } from './report.json';
export { default as pivotReport } from './pivot-report.json';
export { default as webDataStreams } from './webdatastreams.json';
export { default as webDataStreamsBatch } from './webdatastreams-batch.json';
export { default as webDataStreamsBatchSetup } from './webdatastreams-batch-setup.json';
export { default as googleTagSettings } from './google-tag-settings.json';
