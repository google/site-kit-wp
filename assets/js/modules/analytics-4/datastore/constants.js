/**
 * `modules/analytics-4` data store constants.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const MODULES_ANALYTICS_4 = 'modules/analytics-4';

// A special Account ID value used for the "Set up a new account" option.
export const ACCOUNT_CREATE = 'account_create';

// A special Property ID value used for the "Set up a new property" option.
export const PROPERTY_CREATE = 'property_create';

export const WEBDATASTREAM_CREATE = 'webdatastream_create';

export const FORM_SETUP = 'analyticsSetup';

export const MAX_WEBDATASTREAMS_PER_BATCH = 10;

// Date range offset days for Analytics 4 report requests.
export const DATE_RANGE_OFFSET = 1;

export const GTM_SCOPE = 'https://www.googleapis.com/auth/tagmanager.readonly';

export const ENHANCED_MEASUREMENT_FORM = 'enhanced-measurement-form';
export const ENHANCED_MEASUREMENT_ENABLED = 'enhanced-measurement-enabled';
export const ENHANCED_MEASUREMENT_SHOULD_DISMISS_ACTIVATION_BANNER =
	'enhanced-measurement-should-dismiss-activation-banner';

// Form ID for the account creation form.
export const FORM_ACCOUNT_CREATE = 'analyticsAccountCreate';

export const FORM_CUSTOM_DIMENSIONS_CREATE = 'analyticsCustomDimensionsCreate';

// OAuth scope required for provisioning a Google Analytics account.
export const PROVISIONING_SCOPE =
	'https://www.googleapis.com/auth/analytics.provision';
export const EDIT_SCOPE = 'https://www.googleapis.com/auth/analytics.edit';

// Dashboard widget constants.
export const UI_DIMENSION_NAME = 'dashboardAllTrafficWidgetDimensionName';
export const UI_DIMENSION_COLOR = 'dashboardAllTrafficWidgetDimensionColor';
export const UI_DIMENSION_VALUE = 'dashboardAllTrafficWidgetDimensionValue';
export const UI_ACTIVE_ROW_INDEX = 'dashboardAllTrafficWidgetActiveRowIndex';
export const UI_ALL_TRAFFIC_LOADED = 'dashboardAllTrafficWidgetLoaded';

// Note: names and descriptions are not translated as these are not surfaced in Site Kit
// and are also subject to hard limits on the length which would be unpredictable if translated.
// See https://developers.google.com/analytics/devguides/config/admin/v1/rest/v1beta/properties.customDimensions#CustomDimension
export const CUSTOM_DIMENSION_DEFINITIONS = {
	googlesitekit_post_date: {
		parameterName: 'googlesitekit_post_date',
		displayName: 'WordPress Post Date',
		description: 'Created by Site Kit: Date when a post was published',
		scope: 'EVENT',
	},
	googlesitekit_post_author: {
		parameterName: 'googlesitekit_post_author',
		displayName: 'WordPress Post Author',
		description: 'Created by Site Kit: WordPress name of the post author',
		scope: 'EVENT',
	},
	googlesitekit_post_categories: {
		parameterName: 'googlesitekit_post_categories',
		displayName: 'WordPress Post Categories',
		description:
			'Created by Site Kit: Names of categories assigned to a post',
		scope: 'EVENT',
	},
	googlesitekit_post_type: {
		parameterName: 'googlesitekit_post_type',
		displayName: 'WordPress Post Type',
		description: 'Created by Site Kit: Content type of a post',
		scope: 'EVENT',
	},
};

export const ENUM_CONVERSION_EVENTS = {
	ADD_TO_CART: 'add_to_cart',
	PURCHASE: 'purchase',
	SUBMIT_LEAD_FORM: 'submit_lead_form',
	GENERATE_LEAD: 'generate_lead',
	CONTACT: 'contact',
};
export const CONVERSION_REPORTING_LEAD_EVENTS = [
	ENUM_CONVERSION_EVENTS.CONTACT,
	ENUM_CONVERSION_EVENTS.GENERATE_LEAD,
	ENUM_CONVERSION_EVENTS.SUBMIT_LEAD_FORM,
];

// Audience enums.
export const AUDIENCE_FILTER_CLAUSE_TYPE_ENUM = {
	AUDIENCE_CLAUSE_TYPE_UNSPECIFIED: 'AUDIENCE_CLAUSE_TYPE_UNSPECIFIED',
	INCLUDE: 'INCLUDE',
	EXCLUDE: 'EXCLUDE',
};

export const AUDIENCE_FILTER_SCOPE_ENUM = {
	AUDIENCE_FILTER_SCOPE_UNSPECIFIED: 'AUDIENCE_FILTER_SCOPE_UNSPECIFIED',
	AUDIENCE_FILTER_SCOPE_WITHIN_SAME_EVENT:
		'AUDIENCE_FILTER_SCOPE_WITHIN_SAME_EVENT',
	AUDIENCE_FILTER_SCOPE_WITHIN_SAME_SESSION:
		'AUDIENCE_FILTER_SCOPE_WITHIN_SAME_SESSION',
	AUDIENCE_FILTER_SCOPE_ACROSS_ALL_SESSIONS:
		'AUDIENCE_FILTER_SCOPE_ACROSS_ALL_SESSIONS',
};

export const SITE_KIT_AUDIENCE_DEFINITIONS = {
	'new-visitors': {
		description: 'People who visited the site for the first time',
		displayName: 'New visitors',
		membershipDurationDays: -1, // The longest duration, 540 days.
		filterClauses: [
			{
				clauseType: 'INCLUDE',
				simpleFilter: {
					scope: 'AUDIENCE_FILTER_SCOPE_ACROSS_ALL_SESSIONS',
					filterExpression: {
						andGroup: {
							filterExpressions: [
								{
									orGroup: {
										filterExpressions: [
											{
												dimensionOrMetricFilter: {
													fieldName: 'newVsReturning',
													stringFilter: {
														matchType: 'EXACT',
														value: 'new',
													},
												},
											},
										],
									},
								},
								{
									orGroup: {
										filterExpressions: [
											{
												notExpression: {
													dimensionOrMetricFilter: {
														fieldName: 'groupId',
														stringFilter: {
															matchType: 'EXACT',
															value: 'created_by_googlesitekit:new_visitors',
														},
													},
												},
											},
										],
									},
								},
							],
						},
					},
				},
			},
		],
	},
	'returning-visitors': {
		description: 'People who have visited your site at least once before',
		displayName: 'Returning visitors',
		membershipDurationDays: -1, // The longest duration, 540 days.
		filterClauses: [
			{
				clauseType: 'INCLUDE',
				simpleFilter: {
					scope: 'AUDIENCE_FILTER_SCOPE_ACROSS_ALL_SESSIONS',
					filterExpression: {
						andGroup: {
							filterExpressions: [
								{
									orGroup: {
										filterExpressions: [
											{
												dimensionOrMetricFilter: {
													fieldName: 'newVsReturning',
													stringFilter: {
														matchType: 'EXACT',
														value: 'returning',
													},
												},
											},
										],
									},
								},
								{
									orGroup: {
										filterExpressions: [
											{
												notExpression: {
													dimensionOrMetricFilter: {
														fieldName: 'groupId',
														stringFilter: {
															matchType: 'EXACT',
															value: 'created_by_googlesitekit:returning_visitors',
														},
													},
												},
											},
										],
									},
								},
							],
						},
					},
				},
			},
		],
	},
};

export const AUDIENCE_SEGMENTATION_SETUP_FORM = 'audiencePermissionsSetup';

export const AUDIENCE_TILE_CUSTOM_DIMENSION_CREATE =
	'audienceTileCustomDimensionCreate';

export const AUDIENCE_TYPE_SORT_ORDER = {
	USER_AUDIENCE: 0,
	SITE_KIT_AUDIENCE: 1,
	DEFAULT_AUDIENCE: 2,
};

export const AUDIENCE_ITEM_NEW_BADGE_SLUG_PREFIX =
	'audience-selection-panel-expirable-new-badge-';
