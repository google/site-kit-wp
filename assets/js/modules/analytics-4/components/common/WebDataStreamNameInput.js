/**
 * Analytics WebDataStreamNameInput component.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import classnames from 'classnames';
import { useMount } from 'react-use';

/**
 * WordPress dependencies
 */
import { useCallback } from '@wordpress/element';
import { __ } from '@wordpress/i18n';
import { isURL } from '@wordpress/url';

/**
 * Internal dependencies
 */
import { useSelect, useDispatch } from 'googlesitekit-data';
import { CORE_FORMS } from '../../../../googlesitekit/datastore/forms/constants';
import { CORE_SITE } from '../../../../googlesitekit/datastore/site/constants';
import {
	FORM_SETUP,
	MODULES_ANALYTICS_4,
	WEBDATASTREAM_CREATE,
} from '../../datastore/constants';
import AccessibleWarningIcon from '../../../../components/AccessibleWarningIcon';
import { TextField } from 'googlesitekit-components';
import {
	isValidPropertyID,
	isValidWebDataStreamName,
} from '../../utils/validation';

export default function WebDataStreamNameInput() {
	const propertyID = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).getPropertyID()
	);
	const webDataStreamID = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).getWebDataStreamID()
	);
	const webDataStreamName = useSelect( ( select ) =>
		select( CORE_FORMS ).getValue( FORM_SETUP, 'webDataStreamName' )
	);
	const webDataStreamAlreadyExists = useSelect( ( select ) =>
		isValidPropertyID( propertyID )
			? select( MODULES_ANALYTICS_4 ).doesWebDataStreamExist(
					propertyID,
					webDataStreamName
			  )
			: false
	);
	const siteURL = useSelect( ( select ) =>
		select( CORE_SITE ).getReferenceSiteURL()
	);

	const { setValues } = useDispatch( CORE_FORMS );

	const onChange = useCallback(
		( { currentTarget } ) => {
			setValues( FORM_SETUP, { webDataStreamName: currentTarget.value } );
		},
		[ setValues ]
	);

	// Set the default web data stream name.
	useMount( () => {
		if ( ! webDataStreamName && isURL( siteURL ) ) {
			const { hostname } = new URL( siteURL );

			setValues( FORM_SETUP, {
				webDataStreamName: hostname,
			} );
		}
	} );

	// Return early if an existing web data stream is selected.
	if ( webDataStreamID !== WEBDATASTREAM_CREATE ) {
		return null;
	}

	const error =
		webDataStreamAlreadyExists ||
		! webDataStreamName ||
		! isValidWebDataStreamName( webDataStreamName );

	let helperText = false;

	if ( webDataStreamAlreadyExists ) {
		helperText = __(
			'A web data stream with this name already exists.',
			'google-site-kit'
		);
	} else if ( ! webDataStreamName ) {
		helperText = __(
			'A web data stream name is required.',
			'google-site-kit'
		);
	} else if ( ! isValidWebDataStreamName( webDataStreamName ) ) {
		helperText = __(
			'This is not a valid web data stream name.',
			'google-site-kit'
		);
	}

	return (
		<div className="googlesitekit-analytics-webdatastreamname">
			<TextField
				className={ classnames( {
					'mdc-text-field--error': error,
				} ) }
				label={ __( 'Web Data Stream Name', 'google-site-kit' ) }
				outlined
				helperText={ helperText }
				trailingIcon={
					error && (
						<span className="googlesitekit-text-field-icon--error">
							<AccessibleWarningIcon />
						</span>
					)
				}
				value={ webDataStreamName }
				onChange={ onChange }
			/>
		</div>
	);
}
