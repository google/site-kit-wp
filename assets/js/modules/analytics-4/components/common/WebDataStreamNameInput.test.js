/**
 * Analytics WebDataStreamNameInput component tests.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { CORE_FORMS } from '@/js/googlesitekit/datastore/forms/constants';
import { CORE_SITE } from '@/js/googlesitekit/datastore/site/constants';
import {
	FORM_SETUP,
	MODULES_ANALYTICS_4,
	WEBDATASTREAM_CREATE,
} from '@/js/modules/analytics-4/datastore/constants';
import WebDataStreamNameInput from './WebDataStreamNameInput';
import {
	act,
	createTestRegistry,
	render,
} from '../../../../../../tests/js/test-utils';
import * as fixtures from '@/js/modules/analytics-4/datastore/__fixtures__';

describe( 'WebDataStreamNameInput', () => {
	let registry;

	const propertyID = '1001';

	beforeEach( () => {
		registry = createTestRegistry();

		registry.dispatch( CORE_SITE ).receiveSiteInfo( {
			referenceSiteURL: 'http://example.com',
		} );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			propertyID,
		} );

		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetWebDataStreams( fixtures.webDataStreams, {
				propertyID,
			} );

		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.setWebDataStreamID( WEBDATASTREAM_CREATE );
	} );

	it( 'should not render when option to create a new web data stream is not chosen', () => {
		// Select a valid web data stream.
		registry.dispatch( MODULES_ANALYTICS_4 ).setWebDataStreamID( '2001' );

		const { container } = render( <WebDataStreamNameInput />, {
			registry,
		} );

		expect( container ).toBeEmptyDOMElement();
	} );

	it( 'should render when option to create a new web data stream is chosen', () => {
		const { container, getByLabelText } = render(
			<WebDataStreamNameInput />,
			{
				registry,
			}
		);

		expect( getByLabelText( /Web Data Stream Name/i ) ).toBeInTheDocument();

		expect( container ).toMatchSnapshot();
	} );

	it( 'should set value to site URL hostname by default', () => {
		const { getByDisplayValue } = render( <WebDataStreamNameInput />, {
			registry,
		} );

		expect( getByDisplayValue( /example.com/i ) ).toBeInTheDocument();
	} );

	it( 'should show error state if web data stream name already exists', () => {
		// Set the web data stream name to an existing one (in fixtures).
		registry.dispatch( CORE_FORMS ).setValues( FORM_SETUP, {
			webDataStreamName: 'Test GA4 WebDataStream',
		} );

		const { container } = render( <WebDataStreamNameInput />, {
			registry,
		} );

		expect(
			container
				.querySelector( '.mdc-text-field' )
				.classList.contains( 'mdc-text-field--error' )
		).toBeTruthy();

		expect( container ).toHaveTextContent(
			'A web data stream with this name already exists.'
		);

		expect( container ).toMatchSnapshot();
	} );

	it( 'should show error state if web data stream name is not set', () => {
		const { container } = render( <WebDataStreamNameInput />, {
			registry,
		} );

		// Update web data stream name to empty after initial render.
		act( () => {
			registry.dispatch( CORE_FORMS ).setValues( FORM_SETUP, {
				webDataStreamName: '',
			} );
		} );

		expect(
			container
				.querySelector( '.mdc-text-field' )
				.classList.contains( 'mdc-text-field--error' )
		).toBeTruthy();

		expect( container ).toHaveTextContent(
			'A web data stream name is required.'
		);

		expect( container ).toMatchSnapshot();
	} );
} );
