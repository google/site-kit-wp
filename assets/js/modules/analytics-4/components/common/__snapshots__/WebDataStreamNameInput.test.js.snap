// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WebDataStreamNameInput > should render when option to create a new web data stream is chosen 1`] = `
<div>
  <div
    class="googlesitekit-analytics-webdatastreamname"
  >
    <div
      class="mdc-text-field mdc-text-field--outlined"
    >
      <input
        class="mdc-text-field__input"
        id="googlesitekit-textfield-0"
        value="example.com"
      />
      <div
        class="mdc-notched-outline mdc-notched-outline--upgraded mdc-notched-outline--notched"
      >
        <div
          class="mdc-notched-outline__leading"
        />
        <div
          class="mdc-notched-outline__notch"
        >
          <label
            class="mdc-floating-label mdc-floating-label--float-above"
            for="googlesitekit-textfield-0"
            style="transition-duration: 0s;"
          >
            Web Data Stream Name
          </label>
        </div>
        <div
          class="mdc-notched-outline__trailing"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`WebDataStreamNameInput > should show error state if web data stream name already exists 1`] = `
<div>
  <div
    class="googlesitekit-analytics-webdatastreamname"
  >
    <div
      class="mdc-text-field mdc-text-field--error mdc-text-field--outlined mdc-text-field--with-trailing-icon"
    >
      <input
        class="mdc-text-field__input"
        id="googlesitekit-textfield-2"
        value="Test GA4 WebDataStream"
      />
      <div
        class="mdc-notched-outline mdc-notched-outline--upgraded mdc-notched-outline--notched"
      >
        <div
          class="mdc-notched-outline__leading"
        />
        <div
          class="mdc-notched-outline__notch"
        >
          <label
            class="mdc-floating-label mdc-floating-label--float-above"
            for="googlesitekit-textfield-2"
            style="transition-duration: 0s;"
          >
            Web Data Stream Name
          </label>
        </div>
        <div
          class="mdc-notched-outline__trailing"
        />
      </div>
      <span
        class="mdc-text-field__icon googlesitekit-text-field-icon--error"
        tabindex="-1"
      >
        <span
          class="screen-reader-text"
        >
          Error
        </span>
        <svg />
      </span>
    </div>
    <div
      class="mdc-text-field-helper-line"
    >
      <p
        aria-hidden="false"
        class="mdc-text-field-helper-text mdc-text-field-helper-text--persistent"
      >
        A web data stream with this name already exists.
      </p>
    </div>
  </div>
</div>
`;

exports[`WebDataStreamNameInput > should show error state if web data stream name is not set 1`] = `
<div>
  <div
    class="googlesitekit-analytics-webdatastreamname"
  >
    <div
      class="mdc-text-field mdc-text-field--error mdc-text-field--outlined mdc-text-field--with-trailing-icon"
    >
      <input
        class="mdc-text-field__input"
        id="googlesitekit-textfield-3"
        value=""
      />
      <div
        class="mdc-notched-outline mdc-notched-outline--upgraded"
      >
        <div
          class="mdc-notched-outline__leading"
        />
        <div
          class="mdc-notched-outline__notch"
        >
          <label
            class="mdc-floating-label"
            for="googlesitekit-textfield-3"
            style="transition-duration: 0s;"
          >
            Web Data Stream Name
          </label>
        </div>
        <div
          class="mdc-notched-outline__trailing"
        />
      </div>
      <span
        class="mdc-text-field__icon googlesitekit-text-field-icon--error"
        tabindex="-1"
      >
        <span
          class="screen-reader-text"
        >
          Error
        </span>
        <svg />
      </span>
    </div>
    <div
      class="mdc-text-field-helper-line"
    >
      <p
        aria-hidden="false"
        class="mdc-text-field-helper-text mdc-text-field-helper-text--persistent"
      >
        A web data stream name is required.
      </p>
    </div>
  </div>
</div>
`;
